[UnifiedPDF] De-hoist "PDFView" coordinate conversion functions back to PDFPlugin

hortont424 opened the pull request:
#### 47696bacc0e06a2d1d0c63a716961b930f31f0b0
<pre>
[UnifiedPDF] De-hoist &quot;PDFView&quot; coordinate conversion functions back to PDFPlugin
<a href="https://bugs.webkit.org/show_bug.cgi?id=265490">https://bugs.webkit.org/show_bug.cgi?id=265490</a>
<a href="https://rdar.apple.com/118906205">rdar://118906205</a>

Reviewed by Wenson Hsieh and Simon Fraser.

When I hoisted these, I left a FIXME to make sure they make sense.
They don&apos;t. We don&apos;t have any Y flip in UnifiedPDFPlugin.
Move them back to PDFPlugin and add the &quot;PluginToRootView&quot; methods that we really needed.
Also rename frameForHUD to be clear about the coordinate space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::convertFromPluginToPDFView const):
(WebKit::PDFPlugin::convertFromPDFViewToRootView const):
(WebKit::PDFPlugin::convertFromRootViewToPDFView const):
(WebKit::PDFPlugin::convertFromPDFViewToScreen const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::visibilityDidChange):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const):
(WebKit::PDFPluginBase::convertFromPluginToRootView const):
(WebKit::PDFPluginBase::updatePDFHUDLocation):
(WebKit::PDFPluginBase::frameForHUDInRootViewCoordinates const):
(WebKit::PDFPluginBase::convertFromPluginToPDFView const): Deleted.
(WebKit::PDFPluginBase::convertFromPDFViewToRootView const): Deleted.
(WebKit::PDFPluginBase::convertFromRootViewToPDFView const): Deleted.
(WebKit::PDFPluginBase::convertFromPDFViewToScreen const): Deleted.
(WebKit::PDFPluginBase::frameForHUD const): Deleted.

Canonical link: <a href="https://commits.webkit.org/271253@main">https://commits.webkit.org/271253@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a83b021ebccf7d46bb041a2972c08d5b05cc75a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27820 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6457 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30037 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3849 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28085 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/5204 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23855 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4511 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30676 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25408 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25300 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30862 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/6214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6676 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5156 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
lgtm

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1720,6 +1720,41 @@ static void jsPDFDocFinalize(JSObjectRef object)
+
+IntRect PDFPlugin::convertFromPDFViewToRootView(const IntRect& rect) const
+{
+    IntRect rectInPluginCoordinates(rect.x(), rect.y(), rect.width(), rect.height());
+    return valueOrDefault(m_rootViewToPluginTransform.inverse()).mapRect(rectInPluginCoordinates);
```
```suggestion
    return valueOrDefault(m_rootViewToPluginTransform.inverse()).mapRect(rect);
```

Nit - maybe we can just make this a bit cleaner while we're shuffling the code around?

hortont424 replied:
Not going to clean up the to-be-deleted code :)

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1720,6 +1720,41 @@ static void jsPDFDocFinalize(JSObjectRef object)
+
+IntPoint PDFPlugin::convertFromRootViewToPDFView(const IntPoint& point) const
+{
+    IntPoint pointInPluginCoordinates = m_rootViewToPluginTransform.mapPoint(point);
```
I think I want `m_rootViewToPluginTransform` to go away; it's a cache that includes a bunch of ancestor state that surely goes stale in many cases (consider PDF in an <embed> inside a CSS rotation animation).

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1720,6 +1720,41 @@ static void jsPDFDocFinalize(JSObjectRef object)
+
+FloatRect PDFPlugin::convertFromPDFViewToScreen(const FloatRect& rect) const
+{
+    return WebCore::Accessibility::retrieveValueFromMainThread<WebCore::FloatRect>([&] () -> WebCore::FloatRect {
```
This is horrifying but also not new.

hortont424 committed:
[UnifiedPDF] De-hoist "PDFView" coordinate conversion functions back to PDFPlugin
https://bugs.webkit.org/show_bug.cgi?id=265490
rdar://118906205

Reviewed by Wenson Hsieh and Simon Fraser.

When I hoisted these, I left a FIXME to make sure they make sense.
They don't. We don't have any Y flip in UnifiedPDFPlugin.
Move them back to PDFPlugin and add the "PluginToRootView" methods that we really needed.
Also rename frameForHUD to be clear about the coordinate space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::convertFromPluginToPDFView const):
(WebKit::PDFPlugin::convertFromPDFViewToRootView const):
(WebKit::PDFPlugin::convertFromRootViewToPDFView const):
(WebKit::PDFPlugin::convertFromPDFViewToScreen const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::visibilityDidChange):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const):
(WebKit::PDFPluginBase::convertFromPluginToRootView const):
(WebKit::PDFPluginBase::updatePDFHUDLocation):
(WebKit::PDFPluginBase::frameForHUDInRootViewCoordinates const):
(WebKit::PDFPluginBase::convertFromPluginToPDFView const): Deleted.
(WebKit::PDFPluginBase::convertFromPDFViewToRootView const): Deleted.
(WebKit::PDFPluginBase::convertFromRootViewToPDFView const): Deleted.
(WebKit::PDFPluginBase::convertFromPDFViewToScreen const): Deleted.
(WebKit::PDFPluginBase::frameForHUD const): Deleted.

Canonical link: https://commits.webkit.org/271253@main
